### PR TITLE
Regression testing and readability additions for reward curve log2

### PR DESCRIPTION
**Pull request by hoani**
A set of changes to the `log2` function in the `pallet-staking-reward-curve` module.

## Changes

- Add 5 regression tests to check and verify the boundaries of the log2 function
- Readability changes:
  - Pulled `taylor_term` closure out of the `log2` function
  - Add a `pow2!` macro
  - Add doc comments to clarify code intent

## Additional Info

- This PR does not change log2 behaviour

### CODE REVIEW

File: frame/staking/reward-curve/src/log.rs [hunk 1]
```diff
@@ -1,48 +1,69 @@
...
+		}
+		res /= 2 * k as u128 + 1;
+
+		res.try_into().unwrap()
+	}
+}
+
+/// Performs a log2 operation using a rational fraction
+///
+/// result = log2(q/p) where q/p is bound to (0, 1]
```
Comment (thiolliere) [inline frame/staking/reward-curve/src/log.rs:29]: considering `log2(p/q) == -log2(q/p)` I think it is more understandable to say that log2(p, q) == `log2(p/q)` as per-million and with `p>=q`

File: frame/staking/reward-curve/src/lib.rs [hunk 2]
```diff
@@ -268,14 +268,20 @@ impl INPoS {
...
 	}
 
+	// calculates x_ideal from:
+	// y = i_0 + (i_ideal * x_ideal - i_0) * 2^((x_ideal - x)/d)
+	// See web3 docs for the details
 	fn compute_opposite_after_x_ideal(&self, y: u32) -> u32 {
 		if y == self.i_0 {
 			return u32::max_value();
 		}
+		// Note: the log term calculated here represents a negative per_million value
```
Comment (thiolliere) [inline frame/staking/reward-curve/src/lib.rs:278]: considering previous comment this term is not a negative value of `log2((y - self.i_0) / (self.i_ideal_times_x_ideal - self.i_0))` but a positive value of `log2((self.i_ideal_times_x_ideal - self.i_0) /( y - self.i_0));`
```suggestion
		// Note: the log term calculated here represents a per_million value
```

File: frame/staking/reward-curve/src/lib.rs [hunk 3]
```diff
@@ -268,14 +268,20 @@ impl INPoS {
...
 		let log = log2(self.i_ideal_times_x_ideal - self.i_0, y - self.i_0);
 
 		let term: u32 = ((self.d as u64 * log as u64) / 1_000_000).try_into().unwrap();
 
+		// Normally this would be x_ideal - term, but term represents a negative value
+		// because log represents a negative fraction
```
Comment (thiolliere) [inline frame/staking/reward-curve/src/lib.rs:284]:
```suggestion
```

**Comment by hoani** [comment]
Thanks for the clarification on log2(q/p) == -log2(p/q), that makes a lot more sense! I will update asap

File: frame/staking/reward-curve/src/log.rs (hunk 1 above)
Comment (hoani) [reply frame/staking/reward-curve/src/log.rs]: Done, thanks for pointing this out!

**Comment by thiolliere** [comment]
maybe merge master, and also you can change Label when you think it is ready for review again after comments are addressed

**Comment by bkchr** [comment]
Yeah, please merge master.

**Comment by shawntabrizi** [comment]
I don't think I am right to review this PR. Unfamiliar with the math here, would need to look into it better.

That being said, a sanity review of the code looks fine.

I think a test which verifies a range of numbers from the min to max threshold of this function would be nice to see.

**Comment by thiolliere** [comment]
> I don't think I am right to review this PR. Unfamiliar with the math here, would need to look into it better.
> 
> That being said, a sanity review of the code looks fine.
> 
> I think a test which verifies a range of numbers from the min to max threshold of this function would be nice to see.

Does it needs 2 reviewer ? I'm the original writer, and approved, I guess second review should be about substrate 2.0, this code is fine it only adds doc.

also there is already a test for this:
```rust
#[test]
fn test_log() {
	let div = 1_000;
	for p in 0..=div {
		for q in 1..=p {
			let p: u32 = (1_000_000 as u64 * p as u64 / div as u64).try_into().unwrap();
			let q: u32 = (1_000_000 as u64 * q as u64 / div as u64).try_into().unwrap();

			let res = - (log2(p, q) as i64);
			let expected = ((q as f64 / p as f64).log(2.0) * 1_000_000 as f64).round() as i64;
			assert!((res - expected).abs() <= 6);
		}
	}
}
```
